### PR TITLE
feat: 3760/redirect to application

### PR DESCRIPTION
**Pull request by cade-exygy**
# Pull Request Template

## Issue Overview
#3760 #3756

This PR addresses #issue

- [x] This change addresses the issue in full
- [ ] This change addresses only certain aspects of the issue
- [ ] This change is a dependency for another issue
- [ ] This change has a dependency from another issue

## Description
- Redirect after Create Account Confirmation Email
- Redirect after Create Account Resend Confirmation Email
- Redirect after Create Account Timeout
- Redirect after Forget Password Email
- Redirect after Forget Password Timeout
- Edited TypeORM email service

Please include a summary of the change and which issue(s) is addressed. Please also include relevant motivation and context. List any dependencies that are required for this change.

## How Can This Be Tested/Reviewed?
Toggle SHOW_MANDATED_ACCOUNTS var to true
- As a signed out user, do the following:
   - Go to a listing, click apply online, create an account, go to email, click link, and then you should be redirected to application
   - Do the same, but this time hit the 'Resend Confirmation" button in the modal, should work
   - Then if you create an account, hit resend confirmation, and then click on the FIRST email, you should see a timeout modal, which is currently not working on full. I created a bug ticket here: #3846
   - Sign Out, Go to a listing, click apply online, login, redirect to application
   - Sign Out, Go to a listing, click apply online, forget password, email, click link, redirect
   - Toggle the feature toggle off and perform normal sign in/ create flow
   

Describe the tests that you ran to verify your changes. Please also list any relevant details for your test configuration.

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have added QA notes to the issue with applicable URLs
- [x] I have performed a self-review of my own code
- [x] I have reviewed the changes in a desktop view
- [x] I have reviewed the changes in a mobile view
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules
- [ ] I have assigned reviewers
- [ ] I have run `yarn generate:client` and/or created a migration if I made backend changes that require them
- [x] My commit message(s) is/are polished, and any breaking changes are indicated in the message and are well-described
- [ ] Commits made across packages purposefully have the same commit message/version change, else are separated into different commits

## Reviewer Notes:

Steps to review a PR:

- Read and understand the issue, and ensure the author has added QA notes
- Review the code itself from a style point of view
- Pull the changes down locally and test that the acceptance criteria is met
- Also review the acceptance criteria on the Netlify deploy preview (noting that these do not yet include any backend changes made in the PR)
- Either explicitly ask a clarifying question, request changes, or approve the PR if there are small remaining changes but the PR is otherwise good to go

## On Merge:

If you have one commit and message, squash. If you need each message to be applied, rebase and merge.

### CODE REVIEW

File: shared-helpers/src/auth/AuthContext.ts [hunk 1]
```diff
@@ -289,33 +289,45 @@ export const AuthProvider: FunctionComponent<React.PropsWithChildren> = ({ child
...
+    createUser: async (user: UserCreate, listingIdRedirect) => {
       dispatch(startLoading())
+      const appUrl =
+        listingIdRedirect && process.env.showMandatedAccounts
+          ? `${window.location.origin}?redirectUrl=/applications/start/choose-language&listingId=${listingIdRedirect}`
+          : window.location.origin
       try {
         const response = await userService?.create({
-          body: { ...user, appUrl: window.location.origin },
+          body: { ...user, appUrl: appUrl },
```
Comment (emilyjablonski) [inline shared-helpers/src/auth/AuthContext.ts:300]: Nit: In cases where the key/value are the same, you can just do `appUrl`

File: backend/core/src/email/email.service.ts [hunk 2]
```diff
@@ -207,7 +207,15 @@ export class EmailService {
...
     void (await this.loadTranslations(jurisdiction, user.language))
     const compiledTemplate = this.template("forgot-password")
-    const resetUrl = `${appUrl}/reset-password?token=${user.resetToken}`
+    const urlObj = new URL(appUrl)
+
+    const redirectUrl = urlObj.searchParams.get("redirectUrl")
+    const listingId = urlObj.searchParams.get("listingId")
+
+    const resetUrl =
+      redirectUrl && listingId && process.env.SHOW_MANDATED_ACCOUNTS
```
Comment (emilyjablonski) [inline backend/core/src/email/email.service.ts:216]: Nit: You have this pattern in a few places, and it's a very nit comment on performance, but best practice is you generally want to have toggles checked first in a list like this - we don't need to evaluate any logic following a toggle if the toggle is false.

**Review by emilyjablonski — CHANGES_REQUESTED** [review]
Hey Cade - heads up that this isn't working for me. Please let me know if this is working on your end! Following the account creation/email and forgot password/email flows do not redirect me. Potentially the URL is malformed. For example I'm seeing `http://localhost:3000/?redirectUrl=/applications/start/choose-language&listingId=exampleListingId/reset-password?token=exampleBigToken` so I'm never getting to `/reset-password`.

File: backend/core/src/auth/services/user.service.ts [hunk 3]
```diff
@@ -621,7 +621,16 @@ export class UserService {
...
 
   private static getPublicConfirmationUrl(appUrl: string, user: User) {
-    return `${appUrl}?token=${user.confirmationToken}`
+    const urlObj = new URL(appUrl)
+
+    const redirectUrl = urlObj.searchParams.get("redirectUrl")
+    const listingId = urlObj.searchParams.get("listingId")
+
+    const confirmationUrl =
+      redirectUrl && listingId && process.env.SHOW_MANDATED_ACCOUNTS
```
Comment (emilyjablonski) [inline backend/core/src/auth/services/user.service.ts:630]: We don't have access to this in the backend, I'm curious how this was working for you? Did you maybe add it locally to the private env file?

Comment (cade-exygy) [reply backend/core/src/auth/services/user.service.ts:630]: Yes, sorry forgot to mention that. I added the variable in the core env file

Comment (cade-exygy) [reply backend/core/src/auth/services/user.service.ts:630]: I can add it to the template

File: backend/core/.env.template [hunk 4]
```diff
@@ -18,3 +18,4 @@ PARTNERS_PORTAL_URL=http://localhost:3001
 CORS_ORIGINS=["http://localhost:3000", "http://localhost:3001"]
 CORS_REGEX=["test1", "test2"]
 AFS_PROCESSING_CRON_STRING=0 * * * *
+SHOW_MANDATED_ACCOUNTS=FALSE
```
Comment (cade-exygy) [inline backend/core/.env.template:21]: @emilyjablonski @YazeedLoonat @ludtkemorgan Seems like this is the first feature toggle / boolean env variable in the template. How do y'all prefer to handle checking if a env var is true or false? I just did a simple string compare, but happy to change it if you have a preference!

File: sites/public/src/pages/applications/start/choose-language.tsx [hunk 5]
```diff
@@ -55,8 +55,9 @@ const ApplicationChooseLanguage = () => {
     conductor.reset()
     if (!router.isReady && !listingId) return
     if (router.isReady && !listingId) {
-      void router.push("/")
-      return
+      return process.env.showMandatedAccounts && initialStateLoaded && profile
+        ? undefined
```
Comment (cade-exygy) [inline sites/public/src/pages/applications/start/choose-language.tsx:59]: This will be changed in #3769

File: backend/core/.env.template (hunk 4 above)
Comment (cade-exygy) [reply backend/core/.env.template]: Sounds good, just pushed up that change!

File: backend/core/src/auth/services/user.service.ts [hunk 6]
```diff
@@ -621,7 +621,16 @@ export class UserService {
   }
 
   private static getPublicConfirmationUrl(appUrl: string, user: User) {
-    return `${appUrl}?token=${user.confirmationToken}`
+    const urlObj = new URL(appUrl)
```
Comment (YazeedLoonat) [inline backend/core/src/auth/services/user.service.ts:624]: This looks very similar to the one used in the `email.service.ts` file below if you can I would say pull that out into a helper

File: backend/core/src/email/email.service.ts [hunk 7]
```diff
@@ -207,7 +207,15 @@ export class EmailService {
     const jurisdiction = await this.getUserJurisdiction(user)
     void (await this.loadTranslations(jurisdiction, user.language))
     const compiledTemplate = this.template("forgot-password")
-    const resetUrl = `${appUrl}/reset-password?token=${user.resetToken}`
+    const urlObj = new URL(appUrl)
```
Comment (YazeedLoonat) [inline backend/core/src/email/email.service.ts:210]: same as above

File: backend/core/src/auth/services/user.service.ts [hunk 8]
```diff
@@ -621,7 +621,16 @@ export class UserService {
...
+
+    const redirectUrl = urlObj.searchParams.get("redirectUrl")
+    const listingId = urlObj.searchParams.get("listingId")
+
+    const confirmationUrl =
```
Comment (emilyjablonski) [inline backend/core/src/auth/services/user.service.ts:629]: Nit: A suggestion here to reduce repetition / something slightly more readable might just be an if statement that pushes the redirect onto the string (same as below).

**Review by YazeedLoonat — COMMENTED** [review]
I didn't do a full review of this (since I've been so removed from mandated accounts someone else probably has a better read on functionality)

I do have a nitpick on the implementation on the backend though so left a comment about that!

File: shared-helpers/src/auth/AuthContext.ts [hunk 9]
```diff
@@ -289,33 +289,45 @@ export const AuthProvider: FunctionComponent<React.PropsWithChildren> = ({ child
         dispatch(stopLoading())
       }
     },
-    createUser: async (user: UserCreate) => {
+    createUser: async (user: UserCreate, listingIdRedirect) => {
       dispatch(startLoading())
+      const appUrl =
```
Comment (emilyjablonski) [inline shared-helpers/src/auth/AuthContext.ts:294]: Could you reduce the repetition of this block in these three functions? A helper function might help!

File: shared-helpers/src/auth/AuthContext.ts [hunk 10]
```diff
@@ -289,33 +289,45 @@ export const AuthProvider: FunctionComponent<React.PropsWithChildren> = ({ child
...
+    resendConfirmation: async (email: string, listingIdRedirect) => {
       dispatch(startLoading())
+      const appUrl =
+        process.env.showMandatedAccounts && listingIdRedirect
+          ? `${window.location.origin}?redirectUrl=/applications/start/choose-language&listingId=${listingIdRedirect}`
+          : window.location.origin
       try {
         const response = await userService?.resendConfirmation({
-          body: { email, appUrl: window.location.origin },
+          body: { email, appUrl: appUrl },
```
Comment (emilyjablonski) [inline shared-helpers/src/auth/AuthContext.ts:315]: Same comment here from the last round around being able to just have `appUrl` but it's a nit

File: shared-helpers/src/views/sign-in/FormSignIn.tsx [hunk 11]
```diff
@@ -34,7 +35,16 @@ const FormSignIn = ({
     window.scrollTo(0, 0)
   }
   const { LinkComponent } = useContext(NavigationContext)
-
+  const router = useRouter()
+  const listingIdRedirect = router.query?.listingId as string
+  const forgetPasswordURL =
```
Comment (emilyjablonski) [inline shared-helpers/src/views/sign-in/FormSignIn.tsx:40]: This section is a bit hard to read, a suggestion here would be defining the redirect as a single string and having one if statement for both strings that pushes the redirect onto the base.

File: sites/public/src/components/account/ConfirmationModal.tsx [hunk 12]
```diff
@@ -5,6 +5,7 @@ import { useRouter } from "next/router"
 import { useContext, useEffect, useRef, useState } from "react"
 import { useForm } from "react-hook-form"
 import { emailRegex } from "../../lib/helpers"
+import { ParsedUrlQuery } from "querystring"
```
Comment (emilyjablonski) [inline sites/public/src/components/account/ConfirmationModal.tsx:8]: Where is this package pulled in? This type is also just `{ [key: string]: string | string[]; }` - as the type is fairly generic, I don't think we need to pull the package in.

File: sites/public/src/pages/reset-password.tsx [hunk 13]
```diff
@@ -38,13 +39,27 @@ const ResetPassword = () => {
     })
   }, [])
 
+  const handleQueryParams = (query: ParsedUrlQuery) => {
```
Comment (emilyjablonski) [inline sites/public/src/pages/reset-password.tsx:42]: This function is repeated twice and is only called once. I'd recommend just putting this inline, and you may be able to remove the typechecking to return undefined if you do so.

**Review by emilyjablonski — CHANGES_REQUESTED** [review]
Functionally seems great, let me know if you have any questions about the other feedback!

**Review by cade-exygy — COMMENTED** [review]
I think I addressed all of those comments! Thank you for those suggestions!

File: backend/core/src/shared/utils/get-public-email-url.ts [hunk 14]
```diff
@@ -0,0 +1,19 @@
...
+
+  const redirectUrl = urlObj.searchParams.get("redirectUrl")
+  const listingId = urlObj.searchParams.get("listingId")
+
+  let emailUrl = `${urlObj.origin}${urlObj.pathname}/${actionPath ? actionPath : ""}?token=${token}`
+
+  if (!!redirectUrl && !!listingId) {
+    emailUrl = emailUrl.concat(`&redirectUrl=${redirectUrl}&listingId=${listingId}`)
+  }
+  console.log(emailUrl)
```
Comment (emilyjablonski) [inline backend/core/src/shared/utils/get-public-email-url.ts:17]: 🧹

File: backend/core/src/shared/utils/get-public-email-url.ts [hunk 15]
```diff
@@ -0,0 +1,19 @@
+/**
+ * Creates a email URL object from passed url applies redirectUrl and listingId query params if they exist
+ * If they do not exist, the value will be return the email url with just the necessary token
```
Comment (emilyjablonski) [inline backend/core/src/shared/utils/get-public-email-url.ts:3]: `If they do not exist, the return value will be the email url with just the necessary token`

**Review by emilyjablonski — APPROVED** [review]
Note the two comments above, but otherwise lgtm!

**Review by ColinBuyck — APPROVED** [review]
LGTM 🐬 Thanks for the detailed testing instructions

**Pull request merged**